### ISSUE TITLE
Dependencies of Dependencies

### ISSUE THREAD

**Issue by sprankhub** [yireo/Yireo_ExtensionChecker#26]
I think the `PhpClassComponentDetector` is currently taking it a bit too far. `getComponentsByModuleName` gathers all used classed from the module. It then checks each of these classes in `scanClassesForPhpExtensions` for used PHP extensions. So if I inject `\Magento\Framework\View\Page\Config` into my module, I'll get a message that I should add `ext-pcre` to my dependencies, even though I do not use any `preg_*` functions inside my module. Only my dependency `\Magento\Framework\View\Page\Config` uses `preg_replace`. However, we should not report on dependencies of dependencies...?

**Comment by jissereitsma** [comment]
This is a huge issue. I hope to dive into this still today.

**Comment by jissereitsma** [comment]
This took a bit too long. I've released 2.1.2 with the mentioned fix, but will keep this ticket open until an Integration Test will proof this will not break again.

**Comment by jissereitsma** [comment]
This should now be fixed in 2.1.3.

**Comment by sprankhub** [comment]
Thanks!

**Comment by sprankhub** [comment]
I have a very similar issue again. Version 2.2.6 reports a missing dependency on `Magento_Authorization` for a very small module which simply:

* Has a preference on `Magento\AsynchronousOperations\Model\BulkManagement`
* In the preference, it simply "overrides" `\Magento\AsynchronousOperations\Model\BulkManagement::publishOperations` (which won't work, because it is private, but that is another topic - legacy project from another agency :D).

I do not see why this should have a dependency on `Magento_Authorization`. `Magento_Authorization` is only used in the original class, but not in the overridden class at all.

Any idea? Thanks!

**Comment by sprankhub** [comment]
@jissereitsma, could you look into this by chance? I just bumped into this with another example again... We have a block, which extends `\Magento\Backend\Block\Widget`. However, it does NOT override the constructor. But still, the checker finds a dependency on `Magento_Directory`, because the parent constructor `\Magento\Backend\Block\Template::__construct` depends on `\Magento\Directory\Helper\Data`. However, only because a parent class depends on it does not make this a direct dependency of my module.

**Comment by sprankhub** [comment]
@jissereitsma, and again, I bumped into this :D Could you have a look at this, by chance? Thanks and happy holidays!

**Comment by jissereitsma** [comment]
I think I now understand why this popped up. Indeed, with class inheritance, the class inspector of this module simply just reviewed the entire object - not just the child class. I've now commented out those lines, and it fixes things. Version 2.5.0 is now including this fix. Does that work for you?

**Comment by sprankhub** [comment]
@jissereitsma, the release 2.5.0 is still missing in GitHub. Could you check and release? Thanks a lot!

**Comment by sprankhub** [comment]
I think this does the job for us, so that we can close this issue. Since it happened again after some time, it might be worth to write a test for it, but I'm currently not able to contribute it 🙈

**Comment by jissereitsma** [comment]
Yeah, I know - there should be a test for this. I'm behind on this 🕘
